### DISCUSSION TITLE
SQL Server Connection not showing table schemas/columns

### DISCUSSION THREAD

**Discussion by bobpeers** [Budibase/budibase#10115] [Help] [answered]
I have a connection to SQL Server that successfully shows the database tables. However I am unable to either view the table columns or change the permissions to make the data source public.

When I select the table nothing happens and it continues to show the inbuilt DB Users table.

![image](https://user-images.githubusercontent.com/4363445/227861890-39773628-eddf-47af-8798-e835608b4c20.png)

However I have one table in my database that it *is* able to see the schema for. It's the only table without a space in the name but this might be coincidence.
![image](https://user-images.githubusercontent.com/4363445/227862070-005c871d-dd84-4049-b39b-4e47f69bc99f.png)

When previewing the app I *can* see data in the tables but cannot make a public app since there's no way to make the source public. All tables have identical permissions.

**Comment by melohagan** [comment]
Hey @bobpeers 

The space in the table name could possibly be an issue. Would you be able to check the console (Fn+F12 in Chrome) and share screenshots of any errors that you see when viewing one of your SQL tables that does not return data please?

If there's no console errors then check the Network tab for errors.

**Reply by bobpeers** [reply]
Hi @melohagan ,

Thanks for the response. Looking at the Chrome inspector is very interesting. Basically selecting any table always send the request for the SalesdemotestPBI table (if I've previously selected it). Otherwise it returns the built -in users table.
This output is from me selecting 3 different tables, they all produce the same request and response.

![image](https://user-images.githubusercontent.com/4363445/227930601-ee33719a-788d-4e0f-af3e-2fb679d7fd8b.png)

Update - I added the database again and when reading tables I get these errors about no primary key but the tables load without a problem.

![image](https://user-images.githubusercontent.com/4363445/227931848-f5219fbe-58df-4c9c-aded-34e5b8d15a9d.png)

However I also see that all the tables get the sourceID of the last table to load which happens to be the SalesdemotestPBI table. This is why all table are loading the same schema.

![image](https://user-images.githubusercontent.com/4363445/227932240-2f5ec2af-aae8-4e4d-be81-41101062a028.png)

**Reply by melohagan** [reply]
As a test, could you try renaming another table to replace the spaces with underscores to see if that makes a difference?

Or creating a new table with a duplicate schema of a table that isn't working, but with the underscores in the name

**Reply by bobpeers** [reply]
I already have 4 tables without spaces, one has an underscore and the others are just a-z but these don't load at all.

**Reply by bobpeers** [reply]
I've tested just loading one table instead of all and this also doesn't work (even if the table doesn't have spaces in it). It appears that the errors related to no primary key when the schema is retrieved breaks everything after that.
For example I only loaded one table called DateDimension and it can't find the table at all. I also tried loading a table called [NN07 PowerApp Item] which did load but I'm unable to select or view the columns.

**Reply by bobpeers** [reply]
This is trying to load the table DateDimension.

![image](https://user-images.githubusercontent.com/4363445/227938741-7691a88b-c3fa-4896-a071-367817427f36.png)

Result, not loaded

![image](https://user-images.githubusercontent.com/4363445/227938830-1e6bdd99-76eb-4115-8534-14cff7bf5e70.png)

**Reply by melohagan** [reply]
So those tables listed in the error as having *No primary key constraint* won't load because they don't have a primary key.

For tables to function properly in Budibase, they must have a Primary key defined in the SQL Schema.

**Reply by bobpeers** [reply]
Understood and that's fine. The problem is that if I have any tables without a primary key it causes everything else to fail as well. Many tables do have primary keys but these also don't work.
For example the table [NN07 PowerApp Item] has a pk but doesn't work

![image](https://user-images.githubusercontent.com/4363445/227940653-5a48e83b-5857-46e9-996e-eb541f6e99d0.png)

**Reply by melohagan** [reply]
And to double check, if you expand the Constraints folder, do you see a primary key as well? 

Would you be able to share the CREATE script for the table?

**Reply by bobpeers** [reply]
Sure, glad to help :)

![image](https://user-images.githubusercontent.com/4363445/227950476-6e85e806-28f3-4b5a-b0ff-d7ea3532e437.png)


SQL script:

```
SET ANSI_NULLS ON
GO
SET QUOTED_IDENTIFIER ON
GO
CREATE TABLE [dbo].[NN07 PowerApp Item](
	[Master No_] [varchar](255) NOT NULL,
	[Description] [varchar](255) NOT NULL,
	[Description 2] [varchar](255) NOT NULL,
	[EAN] [varchar](255) NOT NULL,
	[Colour Code] [varchar](255) NOT NULL,
	[Colour] [varchar](255) NOT NULL,
	[Starting Date] [datetime] NULL,
	[Ending Date] [datetime] NULL,
	[DKK] [int] NULL,
	[EUR] [int] NULL,
	[GBP] [int] NULL,
	[NOK] [int] NULL,
	[SEK] [int] NULL,
	[AUD] [int] NULL,
	[CAD] [int] NULL,
	[CHF] [int] NULL,
	[USD] [int] NULL
) ON [PRIMARY]
GO
SET ANSI_PADDING ON
GO
ALTER TABLE [dbo].[NN07 PowerApp Item] ADD  CONSTRAINT [pk_powerapp] PRIMARY KEY CLUSTERED 
(
	[EAN] ASC
)WITH (STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ONLINE = OFF, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
GO

```

**Comment by melohagan** [comment]
Hey @bobpeers 

Thank you for the schema. I was able to fetch the table, and insert rows okay. 

![Screenshot 2023-03-28 at 09 35 57](https://user-images.githubusercontent.com/101575380/228178503-14135748-360e-4560-957f-ac6b7b037a19.png)

![Screenshot 2023-03-28 at 09 36 28](https://user-images.githubusercontent.com/101575380/228178631-00f48fec-996b-4edd-9778-6cdbd9be49b5.png)

What was the issue you were seeing?

**Reply by bobpeers** [reply]
I can fetch the table OK

![image](https://user-images.githubusercontent.com/4363445/228188576-1dfe8cfc-490a-49f3-98a0-be5d8ecec0c7.png)

But selecting the table does not show the schema. It just redisplays whatever object was previously selected. in this case the Budibase Users table

![image](https://user-images.githubusercontent.com/4363445/228188870-06471680-32d9-486b-90e7-1dafb0438f5a.png)

**Reply by bobpeers** [reply]
I feel sure it's somehow connected to the errors generated with the tables without a primary key which then breaks the tables that are loaded.
I plan to test this by creating a new database schema with only one table to see if it works. You could also test by creating another table with no primary key and see if you can still see the schema of the PowerApp item table

**Reply by melohagan** [reply]
Thanks @bobpeers that's a good idea. I'll add another table without a primary key to my schema and see if I can reproduce!

**Reply by bobpeers** [reply]
I just tested this by creating a new schema call App with this table as the only table in the schema. The same problem persists. I can load the table with no errors this time about primary keys but still cannot load the schema of the table.

![image](https://user-images.githubusercontent.com/4363445/228224431-acc2acb6-bf0f-44f2-a48e-8b45044b36a9.png)

![image](https://user-images.githubusercontent.com/4363445/228224521-25aba309-a2dc-4368-a146-2798972c2719.png)

I'm stuck what to do now 🤷‍♂️

**Reply by melohagan** [reply]
Hmm that's interesting @bobpeers 

Maybe the app itself has become corrupted. Could you try creating a brand new app and then fetching the SQL Server table in it.

Also which version of Budibase are you running?

**Comment by bobpeers** [comment]
Latest version

![image](https://user-images.githubusercontent.com/4363445/228233013-7ecb8d3f-1b5d-4021-beb8-3aabede122ba.png)

Created a new app but it behaves in exactly the same way as the existing one.

**Comment by bobpeers** [comment]
@melohagan I see you published an update and I can now fetch the schema of all my tables. Many thanks for this, appreciate the quick fix!

**Comment by bobpeers** [comment, marked as answer]
Ahh. Broken again now. Finds the primary key 4 times and can no longer see the table schema like before

![image](https://user-images.githubusercontent.com/4363445/228663400-5cc9e6a3-8761-4fed-bdf5-2579e185199e.png)

**Reply by bobpeers** [reply]
@melohagan Just to add the details it's broken on version 2.4.40 and I tried downgrading to 2.4.39 but that's also broken. I have two app in development and both are broken in the same way, unable to see table schemas. Both apps are using different schemas as well.
Not sure how you fixed it before but something has reverted.

**Reply by melohagan** [reply]
Hey @bobpeers 

Are you cloud or self-hosted?

**Reply by bobpeers** [reply]
I'm on the cloud and I've tried re-adding the server many times to the same result.

**Reply by bobpeers** [reply]
@melohagan I believe you just pushed an update as it works again now :) Here's hoping it stays fixed this time.
